### PR TITLE
[autodiff] [test] Add more for-loop tests for forward mode

### PR DESCRIPTION
**Pull request by erizmr**
Related issue = #5055

### CODE REVIEW

File: tests/python/test_ad_for_fwd.py [hunk 1]
```diff
@@ -212,3 +212,142 @@ def double_for():
...
+    def complex():
+        for i in range(N):
+            weight = 2.0
+            tot = 0.0
+            tot_weight = 0.0
+            for j in range(c[i]):
+                tot_weight += weight + 1
+                tot += (weight + 1) * a[i]
+                weight = weight + 1
+                weight = weight * 4
```
Comment (ailzhang) [inline tests/python/test_ad_for_fwd.py:288]: These might be removed by compiler? maybe use atomic add instead?

**Review by ailzhang — APPROVED** [review]
LGTM!

File: tests/python/test_ad_for_fwd.py (hunk 1 above)
Comment (erizmr) [reply tests/python/test_ad_for_fwd.py:288]: The local atomic add will still be demoted to `load-add-store` before the autodiff pass?

Comment (ailzhang) [reply tests/python/test_ad_for_fwd.py:288]: aha you're right!

**Pull request merged**